### DISCUSSION TITLE
Two Robot Targets

### DISCUSSION THREAD

**Discussion by PhantomStaubli** [visose/Robots#239] [Q&A] [answered]
Robot Targets

Can you please the simplest example?

**Comment by visose** [comment]
You can use simple digital input and output signals for this.
Lets say you want robot A to move to a particular target, and when it reaches that target robot B then starts moving.
Connect a digital output from robot A to a digital input in robot B.
In robot B create a program where a target will have a `WaitDI` command. When you run this program after the robot reaches this target it will start waiting until a signal is received before continuing the program.
In robot A create a program with a target that has a `SetDO` command. When you run this program and the robot reaches that target it will turn on the signal which will cause robot B to start moving.
This is the simplest example.

**Comment by PhantomStaubli** [comment]


**Reply by visose** [reply]
No, it's much more simple than that. This is out of the scope of this plugin, but should be easy to Google.

**Reply by PhantomStaubli** [reply]
I found the physical requirements for the time-based synchronization... 

But for the Robots Plugin to work, do I need to add an "Init Command" to the Input of the "Create Program" component? 

or would I just need to add a "Set DO" component (DO = 1, Value = 1) to the input of the "Command" in the Create Target of Robot A (master/server) and "Wait DI" component (DI = 1) to the input of the "Command" in the Create Target of Robot B (slave/client)? 

Thanks in advance for your time spent. :-)

**Reply by PhantomStaubli** [reply]
I'd appreciate a short update on this. I have had no success running the input commands for the DO and Wait DI. Do I need to modify the XML for the specific robot I am using such as below? If so, how about GH components? I appreciate if you could share the plugin info.


    </Mechanisms>
    <IO>
    <DO names="1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16"/>
    <DI names="1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16"/>
    <AO names="1,2,3,4"/>
    <AI names="1,2,3,4"/>
    </IO></RobotSystem>

</RobotSystems>

**Reply by visose** [reply]
The signal names should match the ones in the controller. See the Bartlett library for examples of Staublis with the default names.

**Comment by PhantomStaubli** [comment]


**Reply by visose** [reply]
The xml data should match the controller data. Don't change the xml if you are not changing the robot base in the controller, etc.

**Comment by PhantomStaubli** [comment]


**Reply by visose** [reply]
There's no `Base` component. Maybe you mean the Base input of the `Load RobotSystem` component. This input is a `Plane` parameter which has a position and rotation.

**Comment by PhantomStaubli** [comment]


**Reply by visose** [reply]
Can you take another picture of the pendant screen expanding the +DInputs and +Douputs? Also share the xml of the robot library you are using. 
My guess is the IO names in the xml are not correct.

**Comment by PhantomStaubli** [comment]


**Reply by PhantomStaubli** [reply]
<RobotSystems>

  <RobotCell name="Bartlett-TX200" manufacturer="Staubli">
    <Mechanisms group="0">
      <RobotArm model="TX200" manufacturer="Staubli" payload="100">
        <Base x="0.000" y="0.000" z="0.000" q1="1.000" q2="0.000" q3="0.000" q4="0.000"/>
        <Joints>
          <Revolute number="1" a ="250" d ="0" minrange = "-180" maxrange ="180" maxspeed ="160"/>
          <Revolute number="2" a ="950" d ="0" minrange = "-115" maxrange ="120" maxspeed ="160"/>
          <Revolute number="3" a ="0" d ="0" minrange = "-140" maxrange ="145" maxspeed ="160"/>
          <Revolute number="4" a ="0" d ="800" minrange = "-270" maxrange ="270" maxspeed ="260"/>
          <Revolute number="5" a ="0" d ="0" minrange = "-120" maxrange ="120" maxspeed ="260"/>
          <Revolute number="6" a ="0" d ="194" minrange = "-270" maxrange ="270" maxspeed ="400"/>
        </Joints>
      </RobotArm>
    </Mechanisms>
    <IO>
      <DO names="BasicIO-1\%Q0,BasicIO-1\%Q1,BasicIO-1\%Q2"/>
      <DI names="BasicIO-1\%I0,BasicIO-1\%I1,BasicIO-1\%I2"/>
    </IO>
  </RobotCell>

</RobotSystems>

**Reply by PhantomStaubli** [reply]
![Screenshot 2023-08-29 at 7 35 49 AM](https://github.com/visose/Robots/assets/137115823/1918ba66-47c3-4fe1-86f7-852ae684e7a7)

**Reply by PhantomStaubli** [reply]
![Screenshot 2023-08-29 at 5 56 39 PM](https://github.com/visose/Robots/assets/137115823/f2ca5973-a8a4-4d7e-a10b-676e15f64d39)
![Uploading Screenshot 2023-08-29 at 5.56.54 PM.png…]()

**Reply by PhantomStaubli** [reply]
![Screenshot 2023-08-29 at 5 56 54 PM](https://github.com/visose/Robots/assets/137115823/99de030a-fe4c-4960-9ef5-95179391d0d9)

**Reply by visose** [reply]
The IO's in the xml seem correct based on the screenshots, but I think the error you posted is from a different program that the one you attached above.

**Reply by visose** [reply]
I think you can also use the name in parenthesis at the right, like `bOut0`
Consider also if the signal is inverted, so false will be true and viceversa.

**Comment by PhantomStaubli** [comment]


**Reply by PhantomStaubli** [reply]
![Screenshot 2023-08-30 at 9 42 45 AM](https://github.com/visose/Robots/assets/137115823/d1f19d07-4ba9-4453-9e8b-7b7d6d98fdfc)
link to the program here. https://www.dropbox.com/scl/fo/szsmh69a9qeoicfjl1953/h?rlkey=q4hpk7q2081oh5zylhcao9ncq&dl=0

**Reply by visose** [reply]
Can you attach the grasshopper file that generates a program that causes the error you posted previously?

**Reply by visose** [reply]
> Now, how do I change the program component to spit out 0-15 instead of the entire IO names (e.g. BasicIO-1\%Q1)?

Edit the xml and just rename the IO names to "0,1,2,3,4". But there was probably a reason it was set up like that.

**Reply by PhantomStaubli** [reply]
Did you have a chance to see the GH? I got stuck here.

**Comment by PhantomStaubli** [comment]


**Reply by visose** [reply]
From your GH file it seems like:
- At the start of robot A, it sets a DO and then waits for a DI.
- At the start of robot B, it waits for a DI and then sets a DO.
This will cause both robots to turn each other on at the very start of the program. You want one robot to wait for DI at the start of the program, but the other robot to turn the DO at the *end* of the program.

**Reply by visose** [reply]
You want the PulseDO command.

**Reply by PhantomStaubli** [reply]
I have not seen PulseDO in Val3...

**Reply by visose** [reply]
Ah, it looks like PuseDO is not implemented for Staubli, you'll need to use SetDO and Wait commands.
For referencing the IOs by index number instead of names, you can edit the XML and add the `useControllerNumbering` attribute to the `IO` element. See: https://github.com/visose/Robots/pull/164 and https://github.com/visose/Robots/blob/libraries/Taiwan%20Tech.xml

**Reply by visose** [reply]
Just add the `useControllerNumbering` attribute, don't delete the DO and DI elements.

**Reply by visose** [reply]
Use the first one you posted here: https://github.com/visose/Robots/discussions/239#discussioncomment-6854021
but just change `<IO>` in line 17 to `<IO useControllerNumbering="true">`
Delete the `<IO>` you added in line 21.

**Reply by visose** [reply]
This plugin out of the box only supports the commands you can find under the `commands` group in the tab bar. There is no command for serial/socket communication. I don't think Grasshopper components are useful for this, it would involve some type of scripting to use in any practical way.

**Reply by visose** [reply]
There would be little value to change the code of this plugin for this. You can write VAL3 code and optionally a separate script (in C#, Python, etc) to communicate with the robot. If you still want to integrate the VAL3 part with the program autogenerated from this plugin you can use the Custom Command and/or Custom Program components.

**Reply by visose** [reply]
If I was you, I would edit the generated code using components to change/replace text to add the links to sio, and feed the modified code to the Custom Code component. Then use the Custom Command to add VAL3 code on specific targets.

**Reply by visose** [reply]
Use the multifile input of the Create Program component to split into more files.

**Reply by PhantomStaubli** [reply]
It will create a separate code file per integers but gives error for Strings. Is there a way to modify the indexes with custom names (e.g. 0_init.pgx, 1_io.pgx...etc)?

**Reply by PhantomStaubli** [reply]
![Screenshot 2023-09-05 at 4 22 32 PM](https://github.com/visose/Robots/assets/137115823/50e07875-53d0-45da-9de5-21b67aa2a93c)

**Comment by visose** [comment, marked as answer]
Closing this long thread with the consolidated answer. For simple two-robot synchronization, use controller IO: one robot sends `SetDO`, the other waits with `WaitDI`, and the IO names in the XML must match the controller or use `useControllerNumbering="true"` on the IO element. PulseDO is not implemented for Staubli, so use SetDO plus Wait commands. Multi-File Indices accept integer split points only; custom output file names are not supported by that input. For socket/serial communication or VAL3-specific setup, use Custom Command/Custom Code or a separate script outside the built-in command set.
